Implement part-pipeline scheme

piotrAMD opened the pull request:
This is a series of commits to implement LGC part-pipeline compilation scheme.

The part-pipeline is guarded by the new option EnablePartPipeline, which
is off by default. It is expected that when all the commits in the series
have been pushed, the old partial-pipeline compilation scheme will go
away.

The part-pipeline scheme is as follows.

First it compiles the FS with FS-applicable pipeline state (or finds in
the shader cache). The FS ELF's PAL metadata contains information on FS
input packing for use when compiling the pre-rasterization
part-pipeline.

Then it compiles the pre-rasterization part-pipeline (VS,TCS,TES,GS)
with pre-rasterization pipeline state and the FS input packing
information (or finds it in the shader cache).

Then it uses the LGC ELF linker to link them together.

See also: [llpc/docs/LlpcOverview.md](llpc/docs/LlpcOverview.md).

Patch by @trenouf with contributions from @bsaleil and myself.

piotrAMD commented:
I realise this PR is quite large, but the way I have split the patch should make it more manageable to review. 

Please expect delayed response time initially, as I am away next week.

s-perron commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+    if (!ok) {
+      errs() << elfLinkerPipeline->getLastError() << "\n";
+      result = Result::ErrorUnavailable;
+      pipelineElf->clear();
```
Should we be clearing at the start of the function?  What if the code fails earlier then the pipeline elf.  We should try to be consistent: if there is an error, then the pipeline elf will be an empty string, or if there is an error then the pipeline elf is undefined.  Either way, we should mention it in the function comments.

piotrAMD replied:
Added a comment on top of the function.

s-perron commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+// @param shaderInfo : Shader info of this graphics pipeline
+// @param [out] pipelineElf : Output Elf package
+// @param [out] stageCacheAccesses : Stage cache access result. All elements must be initialized in the caller as
+//                                   CacheAccessInfo::CacheNotChecked
```
We may have done this for other function, but I do not like it.  If we way this should be initialized to something, then this function should do it.  Otherwise, it should be an [in/out] parameter where we say that an entry is left untouched if the cache it not checked for that entry.

I know this is debatable, so I will not press it if you disagree.

piotrAMD replied:
I agree it would be nicer. This is the same mechanism that is already in use elsewhere, so fixing this properly would require a larger change, which is probably out of scope of this PR (which is quite large already).

s-perron commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+  unsigned wholeStageMask = context->getShaderStageMask();
+  ElfPackage partPipelineBuffers[2];
+  std::unique_ptr<Pipeline> elfLinkerPipeline;
+  std::unique_ptr<ElfLinker> elfLinker;
```
It might be clearer if you just defined an ElfLinker in the stack, and then checked which "partIdx" you were compiling to determine what to do with the elf linker.  I'll add other comments below.

piotrAMD replied:
Reworked this part to check for the stage type instead (didn't go for the stack approach for consistency with other functions).

s-perron commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+    // If we already have an ELF linker pipeline object, we must be doing the pre-rasterization part pipeline,
+    // and the linker's pipeline object contains the FS input mapping state. We need to include that state in
+    // the hash.
+    if (elfLinker) {
```
If all this is really checking is if the pre-rasterization part is being compiled, then actaully check that by checking partIdx.  Then a comment would not be needed.

piotrAMD replied:
Changed to check the stage instead.

s-perron commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+  std::unique_ptr<ElfLinker> elfLinker;
+
+  // For each of the two part pipelines (fragment first, then pre-rasterization)...
+  for (unsigned partIdx = 0; partIdx != 2; ++partIdx) {
```
Could we change this to iterate over an enum?  Then checks like "partIdx ==0" could read `part == Fragment` or `part == PreRasterization`.  At the vert least we should have named constants for these checks.

piotrAMD replied:
Done.

s-perron commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+      }
+    }
+
+    if (!elfLinker) {
```
The code in this block does not need to be in the loop.  I get that it does not have to be executed if the compilation of the fragment part fails, but I expect that to be so rare that it is not worth cluttering the loop.  As I mentioned above, I would remove the use of the unique_ptr, and just make elfLinkerPipeline and elfLinker regular stack objects.  Then all of this could be part of the initialization of these objects at the start of the function.

piotrAMD replied:
Moved this up.

s-perron commented on llpc/context/llpcPipelineContext.cpp:
```diff
@@ -764,11 +838,17 @@ void PipelineContext::setVertexInputDescriptions(Pipeline *pipeline) const {
-void PipelineContext::setColorExportState(Pipeline *pipeline) const {
+// @param [in/out] pipeline : Middle-end pipeline object; nullptr if only hashing
+// @param [in/out] hasher : Hasher object; nullptr if only setting LGC pipeline state
+void PipelineContext::setColorExportState(Pipeline *pipeline, Util::MetroHash64 *hasher) const {
```
I don't like having a single entry point that does two different things.  It should be two functions.  Leave `setColorExportState` create a new function `hashColorExportState`.

trenouf replied:
As above.

s-perron commented on llpc/context/llpcPipelineContext.cpp:
```diff
@@ -615,89 +655,123 @@ void PipelineContext::setUserDataNodesTable(Pipeline *pipeline, ArrayRef<Resourc
-void PipelineContext::setVertexInputDescriptions(Pipeline *pipeline) const {
+// @param [in/out] pipeline : Middle-end pipeline object; nullptr if only hashing
+// @param [in/out] hasher : Hasher object; nullptr if only setting LGC pipeline state
+void PipelineContext::setVertexInputDescriptions(Pipeline *pipeline, Util::MetroHash64 *hasher) const {
```
Same as the color export state.  Two different functions please.

trenouf replied:
I think I did that...
The rationale of the same function is that, when someone adds a new bit of state, it makes it more difficult to forget to add it to the hash.

s-perron replied:
You could have the two entry points that share code behind the scenes.  It is very awkward when you read a line "setVertextInputDescriptions", but it won't actually set anything because pipeline is nullptr.

Also, having the hashing code in PipelineContext is inconsistent with the rest of the hashing code, which is located in its own class.  We should be moving to a single design handling the hashing to make things clearer.  I'm open to which way that is, but I don't want a mixture of designs.

s-perron reviewed (changes requested):
I have a few issues I would like to see fixed up:

1. There are no new tests, but lots of new code.  Tests are needed.  I've commented on some new functionality that absolutely require a test.
2. Some of these changes are useful for than part pipeline compilation.  We should make sure they work in other compilation schemes as well.
3. The intermingling of the hashing code in the pipeline context is awkward.  A function should do 1 thing.  Also, most of the code that does the hashing is found in the poorly named dumper class.  I think it would be useful to all of the hashing code in one class.  It make it easier for people to find.  I find it hard enough when there is a bug caused by something missing from the hash when I know all of the hashing is in 1 file.  If it gets spread out all over the place, it will be that much harder.

kuhar commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+    // See if we want to use part-pipeline compilation, if there is an FS and at least one pre-rasterization
+    // shader stage.
+    unsigned stageMask = context->getShaderStageMask();
+    bool buildPartPipeline = (cl::EnablePartPipeline && (stageMask & (1 << ShaderStageFragment)) &&
+                              (stageMask & ~(1 << ShaderStageFragment)));
```
It would be much better to use helper function to convert enums to/from masks, check if a stage is contained in a mask or not. This is not only more readable but can also be unit-tested for unobvious edge cases.

This also applies to similar code patterns in the rest of the PR.

piotrAMD replied:
Converted to `shaderStageToMask()`.

trenouf commented on lgc/interface/lgc/Pipeline.h:
```diff
@@ -551,6 +551,10 @@ class Pipeline {
   virtual void setShaderStageMask(unsigned mask) = 0;
 
+  // Set whether pre-rasterization part has a geometry shader
+  // NOTE: Only applicable in the part pipeline compilation mode.
```
The difference with "part pipeline compilation" is that you have all the shaders and pipeline state available, you're just choosing to compile the two parts separately. Thus the front-end can tell the LGC compilation of the fragment shader that a GS is present. For general unlinked FS compilation, if the FS does something that relies on knowing whether a GS is present (has 16-bit inputs), then it would have to be one of those cases that fails and falls back to full pipeline compilation.

s-perron replied:
The difference between "part pipeline compilation" and relocatable shader compilation is more just knowledge of the FS inputs.  We should talk about the differences because they could share more code that you think.

In general, LGC should be agnostic to whether or not part pipeline compilation is used or relocatable.  In both cases, it is doing an unlinked compilation.  If the bit is set, then it should be used.

If relocatable shader compilation decides to set this option, which is has to info to do, then it should work.

piotrAMD replied:
This interface change has a very specific use - to do better in/out packing, see the self-contained commit "Part pipeline: Fix for dword packing". Since relocatable shader path does not do the packing, there is no need for relocatable compilation to set this option.

trenouf reviewed (commented):
LGTM modulo Steven's and Jakub's comments. Especially if it passes tests. :-)

piotrAMD commented:
> I have a few issues I would like to see fixed up:
> 
> 1. There are no new tests, but lots of new code.  Tests are needed.  I've commented on some new functionality that absolutely require a test.
> 2. Some of these changes are useful for than part pipeline compilation.  We should make sure they work in other compilation schemes as well.
> 3. The intermingling of the hashing code in the pipeline context is awkward.  A function should do 1 thing.  Also, most of the code that does the hashing is found in the poorly named dumper class.  I think it would be useful to all of the hashing code in one class.  It make it easier for people to find.  I find it hard enough when there is a bug caused by something missing from the hash when I know all of the hashing is in 1 file.  If it gets spread out all over the place, it will be that much harder.

Thanks for the comments. I will add more tests and look into the refactorings suggested.

Re 3, this is a deliberate decision to keep the state and hash in sync, but maybe there is a better way to do that.

s-perron commented:
> Re 3, this is a deliberate decision to keep the state and hash in sync, but maybe there is a better way to do that.

The problem I have is that it creates a separate method for hashing.  Maybe we can try to find a single way to do it.  We can discuss this in another forum.

piotrAMD commented:
Addressing most of the review comments. The changes include:
- A shaderdb test for each of the corner cases where a follow-up fix was needed. In addition to CTS, there is now shaderdb coverage for these cases.
- Some extra tests.
- Using `shaderStageToMask()` instead of bit manipulation.
- Simplifying `setRegister()`.
- Refactoring of `finalizePipeline()`.
- Clang formatting fixes.

The update does not include any particular fixes to the interface commit (Part pipeline: Interface changes), as it needs more work/discussion.

piotrAMD commented on lgc/elfLinker/ElfLinker.cpp:
```diff
@@ -458,6 +458,9 @@ bool ElfLinkerImpl::link(raw_pwrite_stream &outStream) {
       object::ELFSymbolRef elfSymRef(symRef);
+      StringRef name = cantFail(elfSymRef.getName());
+      if (name == "llvmir" && findSymbol(getStringIndex(name)) != 0)
+        continue;
```
It seems the size does not need to be adjusted here. I've added a test.

piotrAMD commented on lgc/patch/PatchResourceCollect.cpp:
```diff
@@ -2535,7 +2535,9 @@ void PatchResourceCollect::updateInputLocInfoMapWithPack() {
-  const bool requireDword = isTcs || isGs || (isFs && m_pipelineState->hasShaderStage(ShaderStageGeometry));
+  bool requireDword = isTcs || isGs ||
+                      (isFs && (m_pipelineState->hasShaderStage(ShaderStageGeometry) ||
+                                (m_pipelineState->isPartPipeline() && (m_pipelineState->getPreRastHasGs()))));
```
Test added.

piotrAMD commented on lgc/include/lgc/state/PalMetadata.h:
```diff
@@ -137,8 +137,9 @@ class PalMetadata {
-  void setRegister(unsigned regNum, unsigned value);
+  // Set a register value in PAL metadata. If "overwrite" is set, the old value is overwritten,
+  // otherwise it is ORed with the new value.
+  void setRegister(unsigned regNum, unsigned value, bool overwrite = false);
```
I've simplified the code so `setRegister()` always overwrite. This should be safe, based on code inspection and some other tests.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -577,6 +577,40 @@ void PalMetadata::finalizePipeline(bool isWholePipeline) {
   if (!isWholePipeline)
     return;
 
+  // Adjust SPI_PS_INPUT_CNTL_0_* if gl_ViewportIndex is not used in the pre-rasterizer stages.
```
I've split the function, plus added tests for the code added.

piotrAMD commented on lgc/elfLinker/ElfLinker.cpp:
```diff
@@ -388,7 +388,9 @@ StringRef ElfLinkerImpl::compileGlue(unsigned glueIndex) {
-  doneInputs();
+  // The call to doneInputs creates any needed glue shaders, but we only need to do it here for unlinked shaders.
+  if (m_pipelineState->isUnlinked())
+    doneInputs();
```
Test added.

amdvlk-admin commented:
### Test summary for commit 3c3d216c9861b225b4a382673fa4a0b72d70def5
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev 9b30a4a91ec444943b23843853dfce2c6618f8fc</li>
<li>XGL: dev 1ce25b1ed8829c27645edd646a3289e4c524c84c</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 0d6cb5407b689f465d37b7f39e2de2c747ef28f1</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/189750)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67723 (56.5%)</li>
<li>Failed: 0/67723 (0.0%)</li>
<li>Not Supported: 29486/67723 (43.5%)</li><li>Warnings: 0/67723 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33884/54304 (62.4%)</li>
<li><details><summary>Failed: 1/54304 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.clustered.graphics.subgroupclusteredmin_uvec3
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 20419/54304 (37.6%)</li><li>Warnings: 0/54304 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67723 (57.9%)</li>
<li>Failed: 0/67723 (0.0%)</li>
<li>Not Supported: 28488/67723 (42.1%)</li><li>Warnings: 0/67723 (0.0%)</li>
</ul></details>


</ul></details>

piotrAMD commented:
Rebased, fixed the sanitizer error and refactored `buildGraphicsPipelineWithPartPipelines` according to Steven's suggestions.

piotrAMD commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1418,12 +1429,162 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+// @param [out] pipelineElf : Output Elf package
+// @param [out] stageCacheAccesses : Stage cache access result. All elements must be initialized in the caller as
+//                                   CacheAccessInfo::CacheNotChecked
+Result Compiler::buildGraphicsPipelineWithPartPipelines(Context *context,
```
Hopefully with the other simplifications, the function's length is now acceptable.

s-perron reviewed (commented):
LGTM

kuhar commented on llpc/test/shaderdb/multiple_inputs/GlslTwoStages.multi-input:
```diff
@@ -2,15 +2,33 @@
+; RUN: | FileCheck -check-prefix=SHADERTEST_PP1 %s
+;
+; SHADERTEST_PP1-LABEL: {{^//}} LLPC final pipeline module info
+; SHADERTEST_PP1:       define dllexport amdgpu_ps { <4 x float> } @_amdgpu_ps_main
+; SHADERTEST_PP1-LABEL: {{^//}} LLPC final pipeline module info
+; SHADERTEST_PP1:       define dllexport amdgpu_vs void @_amdgpu_vs_main
```
What's the difference between this and the previous test? The order of functions in IR?
I'm wondering if we can use the same `SHADERTEST` checks if we use `SHADERTEST-DAG` for both `RUN`s. Also in the other tests.

piotrAMD replied:
Yes, the order of the function is different (and I like the fact that the order is stated explicitly in this test as it shows that there is a difference).

piotrAMD commented:
Rebased, addressed Jakub's comments and disabled the old caching scheme in part-pipeline mode.

piotrAMD commented on lgc/patch/PatchResourceCollect.cpp:
```diff
@@ -2513,7 +2535,9 @@ void PatchResourceCollect::updateInputLocInfoMapWithPack() {
-  const bool requireDword = isTcs || isGs || (isFs && m_pipelineState->hasShaderStage(ShaderStageGeometry));
+  bool requireDword = isTcs || isGs ||
+                      (isFs && (m_pipelineState->hasShaderStage(ShaderStageGeometry) ||
+                                (m_pipelineState->isPartPipeline() && (m_pipelineState->getPreRastHasGs()))));
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+      setUserDataLimit();
+    // If there are root user data nodes but none of them are used, then PAL wants a non-zero user data limit.
+    else if (m_userDataLimit->getUInt() == 0)
+      *m_userDataLimit = 1U;
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+  // Set PA_CL_CLIP_CNTL from pipeline state settings.
+  // DX_CLIP_SPACE_DEF, ZCLIP_NEAR_DISABLE and ZCLIP_FAR_DISABLE are now set internally by PAL (as of
+  // version 629), and are no longer part of the PAL ELF ABI.
+  uint8_t usrClipPlaneMask = m_pipelineState->getRasterizerState().usrClipPlaneMask;
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+  // DX_CLIP_SPACE_DEF, ZCLIP_NEAR_DISABLE and ZCLIP_FAR_DISABLE are now set internally by PAL (as of
+  // version 629), and are no longer part of the PAL ELF ABI.
+  uint8_t usrClipPlaneMask = m_pipelineState->getRasterizerState().usrClipPlaneMask;
+  bool rasterizerDiscardEnable = m_pipelineState->getRasterizerState().rasterizerDiscardEnable;
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+  paClClipCntl.bits.DX_RASTERIZATION_KILL = rasterizerDiscardEnable;
+  setRegister(mmPA_CL_CLIP_CNTL, paClClipCntl.u32All);
+
+  if (m_pipelineState->getTargetInfo().getGfxIpVersion().major >= 9 &&
```
Fixed the gfxip thing. The code was indeed moved but I tried addressing your comments anyway.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+    setRegister(mmDB_SHADER_CONTROL, dbShaderControl.u32All);
+  }
+
+  if (m_pipelineState->getTargetInfo().getGfxIpVersion().major == 10) {
```
I think this is ok, as it specifically checks for the major version.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+  }
+
+  if (m_pipelineState->getTargetInfo().getGfxIpVersion().major == 10) {
+    auto waveBreakSize = m_pipelineState->getShaderOptions(ShaderStageFragment).waveBreakSize;
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -524,60 +614,26 @@ void PalMetadata::finalizePipeline(bool isWholePipeline) {
+  // Finalize register settings.
+  if (m_pipelineState->isGraphics()) {
+    finalizeRegisterSettings(isWholePipeline);
+  }
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+  }
+}
+
+void PalMetadata::finalizeRegisterSettings(bool isWholePipeline) {
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+//
+// Adjust the value if gl_ViewportIndex is not used in the pre-rasterizer stages.
+void PalMetadata::finalizeInputControlRegisterSetting() {
+  assert((m_pipelineState->getShaderStageMask() & shaderStageToMask(ShaderStageFragment)) != 0);
```
I think `isShaderStageInMask` is designed to be used in llpc, not lgc, as it takes `Vkgc::ShaderStage`.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -951,6 +1027,18 @@ bool PalMetadata::fragmentShaderUsesMappedBuiltInInputs() {
   return false;
 }
 
+unsigned PalMetadata::getFragmentShaderBuiltInLoc(unsigned builtin) {
```
Done.

piotrAMD replied:
Done.

piotrAMD commented on llpc/util/llpcUtil.cpp:
```diff
@@ -167,4 +167,13 @@ const char *getUnlinkedShaderStageName(Vkgc::UnlinkedShaderStage type) {
+// @param type : The part-pipeline stage type.
+const char *getPartPipelineStageName(Vkgc::PartPipelineStage type) {
+  static const char *Names[] = {"fragment", "pre-rasterization"};
+  return Names[type];
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+  if (m_pipelineState->getTargetInfo().getGfxIpVersion().major == 10) {
+    auto waveBreakSize = m_pipelineState->getShaderOptions(ShaderStageFragment).waveBreakSize;
+    PA_SC_SHADER_CONTROL paScShaderControl = {};
+    paScShaderControl.gfx10.WAVE_BREAK_REGION_SIZE = static_cast<unsigned int>(waveBreakSize);
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+    bool usesViewportArrayIndex = false;
+    if (m_pipelineNode[Util::Abi::PipelineMetadataKey::UsesViewportArrayIndex].isEmpty())
+      m_pipelineNode[Util::Abi::PipelineMetadataKey::UsesViewportArrayIndex] = false;
+    usesViewportArrayIndex = m_pipelineNode[Util::Abi::PipelineMetadataKey::UsesViewportArrayIndex].getBool();
```
Done.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -951,6 +1027,18 @@ bool PalMetadata::fragmentShaderUsesMappedBuiltInInputs() {
+  auto array2It = m_pipelineNode.find(m_document->getNode(PipelineMetadataKey::FragInputMapping2));
+  if (array2It != m_pipelineNode.end()) {
+    auto fragInputMappingArray2 = array2It->second.getArray(true);
+    for (unsigned idx = 0; idx < fragInputMappingArray2.size() / 2; ++idx) {
```
Done.

piotrAMD commented on lgc/state/ShaderModes.cpp:
```diff
@@ -159,6 +159,16 @@ const ComputeShaderMode &ShaderModes::getComputeShaderMode() {
+// @param stage : Shader stage
+// @param usage : Subgroup size usage
+void ShaderModes::setSubgroupSizeUsage(ShaderStage stage, bool usage) {
+  auto modes = MutableArrayRef<CommonShaderMode>(m_commonShaderModes);
```
Done.

piotrAMD commented on llpc/context/llpcGraphicsContext.cpp:
```diff
@@ -129,15 +129,16 @@ bool GraphicsContext::usesSubgroupSize() const {
-      continue;
-    return true;
+    if (moduleData->usage.useSubgroupSize)
+      bitmask |= (1 << shaderInfoIdx);
```
No - the shaderInfos is declared locally a few lines up.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,97 @@ bool PalMetadata::isGraphics() {
+
+  if (viewportIndexLoc != InvalidValue) {
+
+    bool usesViewportArrayIndex = false;
```
Done.

piotrAMD commented on llpc/context/llpcPipelineContext.h:
```diff
@@ -112,6 +119,12 @@ class PipelineContext {
   // Sets the mask of active shader stages bound to this pipeline
   virtual void setShaderStageMask(unsigned mask) = 0;
 
+  // Sets whether pre-rasterization part has a geometry shader
```
Fixed. 

Just noting here, there are more pre-existing inconsistencies when it comes to a missing dot or using a 3rd person singular ("sets" vs "set").

piotrAMD commented on llpc/context/llpcPipelineContext.h:
```diff
@@ -149,8 +162,8 @@ class PipelineContext {
-  // Set pipeline state in lgc::Pipeline object for middle-end
-  void setPipelineState(lgc::Pipeline *pipeline, bool unlinked) const;
+  // Set pipeline state in lgc::Pipeline object for middle-end, and/or hash the state.
+  void setPipelineState(lgc::Pipeline *pipeline, Util::MetroHash64 *hasher, bool unlinked) const;
```
Done.

piotrAMD commented on llpc/context/llpcGraphicsContext.cpp:
```diff
@@ -129,15 +129,16 @@ bool GraphicsContext::usesSubgroupSize() const {
   // clang-format on
-  for (auto shaderInfo : shaderInfos) {
+  unsigned bitmask = 0;
+  for (unsigned shaderInfoIdx = 0; shaderInfoIdx < shaderInfos.size(); ++shaderInfoIdx) {
```
Done.

piotrAMD commented on llpc/context/llpcPipelineContext.cpp:
```diff
@@ -191,55 +193,72 @@ ShaderHash PipelineContext::getShaderHashCode(ShaderStage stage) const {
-  pipeline->set128BitCacheHash(get128BitCacheHashCode(),
-                               VersionTuple(LLPC_INTERFACE_MAJOR_VERSION, LLPC_INTERFACE_MINOR_VERSION));
+  for (ShaderStage stage : make_filter_range(stages, isNativeStage)) {
+    if (stageMask & (1 << stage))
```
Done.

piotrAMD commented on llpc/context/llpcPipelineContext.cpp:
```diff
@@ -615,89 +655,123 @@ void PipelineContext::setUserDataNodesTable(Pipeline *pipeline, ArrayRef<Resourc
+    rasterizerState.provokingVertexMode = static_cast<ProvokingVertexMode>(inputRsState.provokingVertexMode);
+  }
+
+  if (stageMask & shaderStageToMask(ShaderStageFragment)) {
```
Done.

piotrAMD commented on llpc/context/llpcPipelineContext.cpp:
```diff
@@ -615,89 +655,123 @@ void PipelineContext::setUserDataNodesTable(Pipeline *pipeline, ArrayRef<Resourc
   }
 
-  pipeline->setDepthStencilState(depthStencilState);
+  if (stageMask & shaderStageToMask(ShaderStageFragment)) {
```
Done.

piotrAMD commented on llpc/test/shaderdb/general/PipelineVsFs_VsWithData.pipe:
```diff
@@ -22,25 +22,47 @@
-; SHADERTEST2-NEXT: {{[0-9]*}}: 00000000
-; SHADERTEST2-NEXT: {{[0-9]*}}: 3F800000
-; SHADERTEST2-NEXT: {{[0-9]*}}: 3F800000
+; RUN: amdllpc -enable-part-pipeline=0 -spvgen-dir=%spvgendir% -o %t.elf %gfxip %s && llvm-objdump --arch=amdgcn --disassemble-zeroes --mcpu=gfx900 -d -j .text -j .rodata -r %t.elf | FileCheck -check-prefix=SHADERTEST2_PP0 %s
```
Done.

piotrAMD commented on llpc/test/shaderdb/general/PipelineVsFs_VsWithData.pipe:
```diff
@@ -22,25 +22,47 @@
+; END_SHADERTEST
+
+; BEGIN_SHADERTEST
+; RUN: amdllpc -enable-part-pipeline=1 -spvgen-dir=%spvgendir% -o %t.elf %gfxip %s && llvm-objdump --arch=amdgcn --disassemble-zeroes --mcpu=gfx900 -d -j .text -j .rodata -r %t.elf | FileCheck -check-prefix=SHADERTEST2_PP1 %s
```
Done.

piotrAMD commented on llpc/test/shaderdb/gfx10/PipelineVsFs_TestSubgroupSizeUsageFragment.pipe:
```diff
@@ -0,0 +1,29 @@
+; RUN: amdllpc -enable-part-pipeline=0 -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; RUN: amdllpc -enable-part-pipeline=1 -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; SHADERTEST-LABEL: LLPC final ELF info
+; SHADERTEST: VGT_SHADER_STAGES_EN                          0x0000000000010000
```
Done.

piotrAMD commented on llpc/test/shaderdb/gfx10/PipelineVsFs_TestSubgroupSizeUsageVertex.pipe:
```diff
@@ -0,0 +1,33 @@
+; RUN: amdllpc -enable-part-pipeline=0 -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; RUN: amdllpc -enable-part-pipeline=1 -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; SHADERTEST-LABEL: LLPC final ELF info
+; SHADERTEST: VGT_SHADER_STAGES_EN                          0x0000000000010000
```
Done.

piotrAMD commented on llpc/test/shaderdb/gfx9/PipelineVsFs_TestAlpha2Coverage.pipe:
```diff
@@ -0,0 +1,29 @@
+; RUN: amdllpc -enable-part-pipeline=0 -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; RUN: amdllpc -enable-part-pipeline=1 -spvgen-dir=%spvgendir% -v %gfxip %s | FileCheck -check-prefix=SHADERTEST %s
+; SHADERTEST-LABEL: LLPC final ELF info
+; SHADERTEST: DB_SHADER_CONTROL                             0x0000000000000010
```
Done.

amdvlk-admin commented:
### Test summary for commit 3d34f7a0fd07a0d53092ebaa2847d2c9470c4663
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202608)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67536 (56.6%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29299/67536 (43.4%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67536 (56.4%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29461/67536 (43.6%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67536 (58.1%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 28301/67536 (41.9%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>


</ul></details>

s-perron reviewed (dismissed):
LGTM.  It think everything has been dealt with for now.

kuhar commented on llpc/context/llpcPipelineContext.h:
```diff
@@ -97,8 +103,9 @@ class PipelineContext {
-  // Check whether the pipeline uses features relevant to subgroup size
-  virtual bool usesSubgroupSize() const = 0;
+
+  // Gets subgroup size usage - bitmask with one bit per stage.
```
This does not  say anything about the order of the bits. Is that important?

piotrAMD replied:
It's not important for the users of the function right now, but I've extended the comment.

piotrAMD commented on lgc/state/PalMetadata.cpp:
```diff
@@ -524,60 +614,26 @@ void PalMetadata::finalizePipeline(bool isWholePipeline) {
     return;
 
+  // In the part-pipeline compilation only at ELF link stage do we know how gl_ViewportIndex was used in all stages.
+  if ((m_pipelineState->getShaderStageMask() & shaderStageToMask(ShaderStageFragment)) != 0) {
```
Done.

piotrAMD commented on llpc/context/llpcPipelineContext.cpp:
```diff
@@ -420,9 +445,24 @@ void PipelineContext::setOptionsInPipeline(Pipeline *pipeline) const {
+    // TODO: Improve the API here to let us pass the mask.
+    ShaderStage userDataStage = ShaderStageInvalid; // All shader stages
+    if (llvm::isPowerOf2_32(stageMask))
+      userDataStage = ShaderStage(llvm::Log2_32(stageMask));
```
Done.

piotrAMD commented on lgc/include/lgc/state/AbiMetadata.h:
```diff
@@ -366,4 +367,24 @@ union PA_SC_AA_CONFIG {
 
+union SPI_PS_INPUT_CNTL_0 {
+  struct {
+    unsigned int OFFSET : 6;
```
Done.

piotrAMD commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1230,12 +1243,156 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+    MetroHash::Hash partPipelineHash = {};
+    hasher.Finalize(partPipelineHash.bytes);
+    llvm::Optional<CacheAccessor> cacheAccessor;
+    cacheAccessor.emplace(context, partPipelineHash, getInternalCaches());
```
Done.

piotrAMD commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1230,12 +1243,156 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+    // Hash the selected shaders and the pipeline state applicable to them.
+    Util::MetroHash64 hasher;
+    for (unsigned shaderIndex = 0; shaderIndex < shaderInfo.size() && result == Result::Success; ++shaderIndex) {
+      if (const PipelineShaderInfo *shaderInfoEntry = shaderInfo[shaderIndex]) {
```
Done.

piotrAMD commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1230,12 +1243,156 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+      // Keep the PipelineShaderInfo structs for the shaders we are including.
+      llvm::SmallVector<const PipelineShaderInfo *, 4> partPipelineShaderInfo;
+      for (const PipelineShaderInfo *oneShaderInfo : shaderInfo) {
+        if (oneShaderInfo && (partStageMask & shaderStageToMask(oneShaderInfo->entryStage)))
```
Done - added a new helper in lgc.

piotrAMD commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1230,12 +1243,156 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+// @param [out] pipelineElf : Output Elf package. If an error occurs, the content is undefined.
+// @param [out] stageCacheAccesses : Stage cache access result. All elements must be initialized in the caller as
+//                                   CacheAccessInfo::CacheNotChecked
+Result Compiler::buildGraphicsPipelineWithPartPipelines(Context *context,
```
Can we defer it until the PR with hashing refactor? The hashing part is a good candidate for a piece of code to be extracted from this function.

kuhar replied:
SGTM

piotrAMD commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1230,12 +1243,156 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+    elfLinker->addInputElf(llvm::MemoryBufferRef(partPipelineElf, ""));
+  }
+
+  if (result == Result::Success) {
```
Done.

piotrAMD commented on llpc/context/llpcComputeContext.h:
```diff
@@ -56,11 +56,14 @@ class ComputeContext : public PipelineContext {
   // Sets the mask of active shader stages bound to this pipeline
   void setShaderStageMask(unsigned mask) { assert(mask == getShaderStageMask()); }
 
+  void setPreRastHasGs(bool preRastHasGs) { llvm_unreachable("Should never be called!"); }
```
Done.

piotrAMD commented on llpc/context/llpcGraphicsContext.cpp:
```diff
@@ -118,8 +118,8 @@ const PipelineShaderInfo *GraphicsContext::getPipelineShaderInfo(ShaderStage sha
 // =====================================================================================================================
-// Check whether the pipeline uses features relevant to subgroup size
-bool GraphicsContext::usesSubgroupSize() const {
+// Gets subgroup size usage
```
Done.

piotrAMD commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1230,12 +1243,156 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+      // Mark the applicable entries in stageCacheAccesses.
+      for (unsigned shaderStage = 1 << llvm::countTrailingZeros(partStageMask); 1 << shaderStage <= partStageMask;
+           ++shaderStage) {
+        if (partStageMask & shaderStageToMask(static_cast<ShaderStage>(shaderStage))) {
```
Done.

piotrAMD commented:
Addressed recent Jakub's comments. 
As for clang-tidy warnings, I added NOLINTBEGIN/NOLINTEND (introduced in clang-tidy 14) but they do not appear to work.

amdvlk-admin commented:
### Test summary for commit 7686a336d2ab7f2190062e2a041160886dad0535
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202431)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67477 (56.7%)</li>
<li>Failed: 0/67477 (0.0%)</li>
<li>Not Supported: 29240/67477 (43.3%)</li><li>Warnings: 0/67477 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67477 (56.4%)</li>
<li>Failed: 0/67477 (0.0%)</li>
<li>Not Supported: 29402/67477 (43.6%)</li><li>Warnings: 0/67477 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67477 (58.1%)</li>
<li>Failed: 0/67477 (0.0%)</li>
<li>Not Supported: 28242/67477 (41.9%)</li><li>Warnings: 0/67477 (0.0%)</li>
</ul></details>


</ul></details>

kuhar commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -1230,12 +1243,153 @@ Result Compiler::buildGraphicsPipelineInternal(GraphicsContext *graphicsContext,
+      auto otherPartPipeline = (partPipelineStage == PartPipelineStagePreRasterization) ? &*elfLinkerPipeline : nullptr;
+      result = buildPipelineInternal(context, partPipelineShaderInfo, PipelineLink::PartPipeline, otherPartPipeline,
+                                     &partPipelineBuffers[partPipelineStage], stageCacheAccesses);
+      if (result != Result::Success)
+        break;
```
nit: can we `return` here instead of `break`ing? If yes, we could get rid of the `result` variable.

kuhar reviewed (approved):
LGTM modulo a few outstanding things.

---

To disable most clang-tidy warnings:
> To disable a lint, add `// NOLINT` at the end of the line.

I skimmed the list of warnings and this won't work in the case of the first function definition because this is a compiler warning and not a lint warning proper. In that case remove `static` since `inline` is sufficient.

piotrAMD commented:
> LGTM modulo a few outstanding things.
> 
> To disable most clang-tidy warnings:
> 
> > To disable a lint, add `// NOLINT` at the end of the line.
> 
> I skimmed the list of warnings and this won't work in the case of the first function definition because this is a compiler warning and not a lint warning proper. In that case remove `static` since `inline` is sufficient.

I think you're referring to an issue in initializeStatePasses in PipelineState.h? This is not introduced by this patch.

I was trying disabling invalid cast warning in AbiMetadata.h, but BEGIN/END tokens do not appear to work (I suspect an old clang-tidy is used - pre v14).

kuhar commented:
> I think you're referring to an issue in initializeStatePasses in PipelineState.h? This is not introduced by this patch.

Ack. We should fix that separately then.

> I was trying disabling invalid cast warning in AbiMetadata.h, but BEGIN/END tokens do not appear to work (I suspect an old clang-tidy is used - pre v14).

Can you use per-line tokens instead? (`//NOLINT`) Otherwise it will keep complaining to everyone.

piotrAMD commented:
Simplified result code handling in the interface and added NOLINT in the new lines in `AbiMetadata.h`.

amdrexu commented on include/vkgcDefs.h:
```diff
@@ -276,6 +276,13 @@ enum class ResourceMappingNodeType : unsigned {
   Count, ///< Count of resource mapping node types.
 };
 
+/// Enumerates part-pipeline stages of compilation.
```
Could bump the interface minor version

amdrexu commented on include/vkgcDefs.h:
```diff
@@ -276,6 +276,13 @@ enum class ResourceMappingNodeType : unsigned {
 
+/// Enumerates part-pipeline stages of compilation.
+enum PartPipelineStage : unsigned {
+  PartPipelineStageFragment,
```
Better to add comments for each enumerants.

amdrexu commented on lgc/include/lgc/state/PalMetadata.h:
```diff
@@ -92,6 +92,9 @@ struct FsInputMappings {
   // such as CullDistance and ClipDistance, the built-in id and the mapped location number.
   llvm::SmallVector<std::pair<unsigned, unsigned>> builtInLocationInfo;
+  // Array sizes for ClipDistance and CullDistance.
+  unsigned clipDistance;
```
-> clipDistanceCount

amdrexu commented on lgc/include/lgc/state/PalMetadata.h:
```diff
@@ -92,6 +92,9 @@ struct FsInputMappings {
   llvm::SmallVector<std::pair<unsigned, unsigned>> builtInLocationInfo;
+  // Array sizes for ClipDistance and CullDistance.
+  unsigned clipDistance;
+  unsigned cullDistance;
```
cullDistance -> cullDistanceCount

amdrexu commented on lgc/include/lgc/state/PipelineState.h:
```diff
@@ -132,6 +132,10 @@ class PipelineState final : public Pipeline {
 
+  // Set whether pre-rasterization part has a geometry shader
+  // NOTE: Only applicable in the part pipeline compilation mode.
+  void setPreRastHasGs(bool preRastHasGs) override final { m_preRastHasGs = preRastHasGs; }
```
Prefer to use Raster as the abbreviation
setPreRastHasGs -> setPreRasterHasGs
getPreRastHasGs -> getPreRasterHasGs

amdrexu commented on lgc/include/lgc/state/PipelineState.h:
```diff
@@ -475,6 +480,7 @@ class PipelineState final : public Pipeline {
   // Whether generating pipeline or unlinked part-pipeline
   PipelineLink m_pipelineLink = PipelineLink::WholePipeline;
   unsigned m_stageMask = 0;                             // Mask of active shader stages
+  bool m_preRastHasGs = false;                          // Whether pre-rasterization part has a geometry shader
```
m_preRastHasGs -> m_preRasterHasGs

amdrexu commented on lgc/include/lgc/state/ShaderStage.h:
```diff
@@ -52,6 +52,16 @@ constexpr unsigned shaderStageToMask(Stage theStage, Stages... otherStages) {
+//
+// @param stage : Shader stage to look for
+// @param stageMask : Stage mask to check
+// @returns : true iff `stageMask` contains `stage`
```
true -> True

amdrexu commented on lgc/interface/lgc/Pipeline.h:
```diff
@@ -551,6 +551,10 @@ class Pipeline {
 
+  // Set whether pre-rasterization part has a geometry shader
+  // NOTE: Only applicable in the part pipeline compilation mode.
+  virtual void setPreRastHasGs(bool preRastHasGs) = 0;
```
setPreRastHasGs -> setPreRasterHasGs

amdrexu commented on lgc/state/PalMetadata.cpp:
```diff
@@ -506,13 +512,100 @@ bool PalMetadata::isGraphics() {
+}
+
+// =====================================================================================================================
+// Finalize PAL register settings for pipeline, part-pipeline or shader compilation.
```
Missing the parameter comment `isWholePipeline`

amdrexu commented on lgc/state/PalMetadata.cpp:
```diff
@@ -951,6 +1028,20 @@ bool PalMetadata::fragmentShaderUsesMappedBuiltInInputs() {
 }
 
+// =====================================================================================================================
+// Returns the location of the fragment builtin or InvalidValue if the builtin is not found.
```
Missing parameter comment of `builtin`

amdrexu commented on lgc/state/PipelineState.cpp:
```diff
@@ -52,6 +52,7 @@ static cl::opt<bool> EnableTessOffChip("enable-tess-offchip", cl::desc("Enable t
 
 // Names for named metadata nodes when storing and reading back pipeline state
 static const char UnlinkedMetadataName[] = "lgc.unlinked";
+static const char PreRastHasGsMetadataName[] = "lgc.prerast.has.gs";
```
PreRastHasGsMetadataName -> PreRasterHasGsMetadataName

amdrexu commented on llpc/context/llpcCompiler.cpp:
```diff
@@ -816,11 +820,14 @@ bool Compiler::canUseRelocatableComputeShaderElf(const ComputePipelineBuildInfo
-// @param unlinked : Do not provide some state to LGC, so offsets are generated as relocs
+// @param pipelineLink : WholePipeline = whole pipeline compile
+//                       Unlinked = shader or part-pipeline compiled without pipeline state such as vertex fetch
+// @param otherPartPipeline : nullptr or the other Pipeline object containing FS input mappings for a part-pipeline
```
nullptr -> Nullptr

amdrexu commented on llpc/context/llpcGraphicsContext.h:
```diff
@@ -58,14 +58,18 @@ class GraphicsContext : public PipelineContext {
   void setShaderStageMask(unsigned mask) { m_stageMask = mask; }
 
+  // Gets whether pre-rasterization part has a geometry shader
+  void setPreRastHasGs(bool preRastHasGs) { m_preRastHasGs = preRastHasGs; }
```
setPreRastHasGs -> setPreRasterHasGs

amdrexu commented on llpc/context/llpcGraphicsContext.h:
```diff
@@ -58,14 +58,18 @@ class GraphicsContext : public PipelineContext {
 
+  // Gets whether pre-rasterization part has a geometry shader
+  void setPreRastHasGs(bool preRastHasGs) { m_preRastHasGs = preRastHasGs; }
+  bool getPreRastHasGs() const { return m_preRastHasGs; };
```
The same as above

amdrexu commented on llpc/context/llpcGraphicsContext.h:
```diff
@@ -75,6 +79,7 @@ class GraphicsContext : public PipelineContext {
   const GraphicsPipelineBuildInfo *m_pipelineInfo; // Info to build a graphics pipeline
 
   unsigned m_stageMask;        // Mask of active shader stages bound to this graphics pipeline
+  bool m_preRastHasGs;         // Whether pre-rasterization part has a geometry shader
```
m_preRastHasGs -> m_preRasterHasGs

amdrexu commented on llpc/context/llpcPipelineContext.h:
```diff
@@ -112,6 +120,14 @@ class PipelineContext {
 
+  // Sets whether pre-rasterization part has a geometry shader.
+  // NOTE: Only applicable in the part pipeline compilation mode.
+  virtual void setPreRastHasGs(bool preRastHasGs) = 0;
```
setPreRastHasGs -> setPreRasterHasGs

amdrexu commented on llpc/context/llpcPipelineContext.h:
```diff
@@ -112,6 +120,14 @@ class PipelineContext {
+
+  // Gets whether pre-rasterization part has a geometry shader.
+  // NOTE: Only applicable in the part pipeline compilation mode.
+  virtual bool getPreRastHasGs() const = 0;
```
The same as above

amdrexu reviewed (commented):
Just some naming nits. LGTM.

JaxLinAMD commented:
retest this please

JaxLinAMD commented:
retest this please

amdvlk-admin commented:
### Test summary for commit dd24db7a97e9c717160666c6e52305f85c8ae64c
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 36c5637649089bf65d42c36b431f47c170b57276</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev eb9d3792bdfda0b6fb94af6bd9989cad51842403</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/189214)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67464 (56.7%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 29227/67464 (43.3%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33866/54286 (62.4%)</li>
<li><details><summary>Failed: 1/54286 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.clustered.graphics.subgroupclusteredadd_int8_t
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 20419/54286 (37.6%)</li><li>Warnings: 0/54286 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67464 (58.2%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 28229/67464 (41.8%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>


</ul></details>

piotrAMD commented:
Addressed Rex's comments.

amdvlk-admin commented:
### Test summary for commit 4113c4ceff0a9801700d873e6d1fbecbcd7d9e0d
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 36c5637649089bf65d42c36b431f47c170b57276</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev eb9d3792bdfda0b6fb94af6bd9989cad51842403</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/201585)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38010/67195 (56.6%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 29185/67195 (43.4%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 37848/67195 (56.3%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 29347/67195 (43.7%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39069/67195 (58.1%)</li>
<li>Failed: 0/67195 (0.0%)</li>
<li>Not Supported: 28126/67195 (41.9%)</li><li>Warnings: 0/67195 (0.0%)</li>
</ul></details>


</ul></details>
